File: src/test/java/pages/ContactCreationPage.java
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import io.qameta.allure.Step;

public class ContactCreationPage {
    private WebDriver driver;

    // Locators
    public By firstNameInput = By.name("first_name");
    public By lastNameInput = By.name("last_name");
    public By emailAddressInput = By.xpath("//input[@placeholder='Email address']");
    public By emailTypeInput = By.xpath("//input[@placeholder='Personal email, Business, Alt...']");
    public By addEmailButton = By.cssSelector("button.ui.tiny.basic.icon.button i.add.icon");
    public By categoryDropdown = By.cssSelector("div[name='category']");
    public By saveButton = By.cssSelector("button.ui.linkedin.button i.save.icon");
    public By confirmationMessage = By.xpath("//div[contains(@class,'confirmation') or contains(text(),'Contact saved')]"); // TODO: Replace with actual locator

    // Constructor
    public ContactCreationPage(WebDriver driver) {
        this.driver = driver;
    }

    /**
     * Complete the contact creation flow with maximum characters in the Name field.
     * @param maxFirstName String with maximum allowed characters for First Name
     * @param lastName Valid last name
     * @param email Valid email address
     * @param emailType Email type (e.g., Personal, Business)
     * @param category Category to select (e.g., Lead, Customer)
     */
    @Step("Create contact with maximum character limit for Name field")
    public void createContactWithMaxName(String maxFirstName, String lastName, String email, String emailType, String category) {
        // Enter First Name (max chars)
        driver.findElement(firstNameInput).clear();
        driver.findElement(firstNameInput).sendKeys(maxFirstName);

        // Enter Last Name
        driver.findElement(lastNameInput).clear();
        driver.findElement(lastNameInput).sendKeys(lastName);

        // Enter Email Address
        driver.findElement(emailAddressInput).clear();
        driver.findElement(emailAddressInput).sendKeys(email);

        // Enter Email Type
        driver.findElement(emailTypeInput).clear();
        driver.findElement(emailTypeInput).sendKeys(emailType);

        // Click Add Email Button
        driver.findElement(addEmailButton).click();

        // Select Category from Dropdown
        driver.findElement(categoryDropdown).click();
        // Select the category option (e.g., Lead, Customer, etc.)
        WebElement categoryOption = driver.findElement(By.xpath("//div[@name='category']//span[text()='" + category + "']"));
        categoryOption.click();

        // Click Save Button
        driver.findElement(saveButton).click();
    }

    /**
     * Check if confirmation message is displayed after saving contact.
     * @return true if confirmation message is displayed, false otherwise
     */
    public boolean isConfirmationMessageDisplayed() {
        try {
            return driver.findElement(confirmationMessage).isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Fills the contact creation form with maximum allowed characters in the Name field and saves the contact.
     * This method is specifically for test case TC-N004: Maximum Character Limit for Name Field.
     *
     * @param maxFirstName String with maximum allowed characters for First Name
     * @param lastName Valid last name
     * @param email Valid email address
     * @param company Company name
     * @param position Position or job title
     */
    public void createContactWithMaxNameField(String maxFirstName, String lastName, String email, String company, String position) {
        // Enter First Name (max chars)
        driver.findElement(firstNameInput).clear();
        driver.findElement(firstNameInput).sendKeys(maxFirstName);

        // Enter Last Name
        driver.findElement(lastNameInput).clear();
        driver.findElement(lastNameInput).sendKeys(lastName);

        // Enter Email Address
        driver.findElement(emailAddressInput).clear();
        driver.findElement(emailAddressInput).sendKeys(email);

        // Enter Company
        By companyDropdown = By.cssSelector("div[role='combobox']");
        driver.findElement(companyDropdown).click();
        // Enter company name in dropdown search input
        By companySearchInput = By.cssSelector("div[role='combobox'] input.search");
        driver.findElement(companySearchInput).sendKeys(company);
        // Select the company from the dropdown (assumes exact match)
        By companyOption = By.xpath("//div[@role='combobox']//span[text()='" + company + "']");
        if (driver.findElements(companyOption).size() > 0) {
            driver.findElement(companyOption).click();
        }

        // Enter Position (assuming a placeholder locator as it's not defined)
        By positionInput = By.xpath("<PLACEHOLDER_Position_Input>"); // TODO: Replace with actual locator
        if (driver.findElements(positionInput).size() > 0) {
            driver.findElement(positionInput).clear();
            driver.findElement(positionInput).sendKeys(position);
        }

        // Click Save Button
        By saveButton = By.xpath("//button[contains(@class, 'linkedin button') and contains(text(), 'Save')]");
        driver.findElement(saveButton).click();
    }

    /**
     * Checks if the confirmation message is displayed after saving the contact.
     * @return true if confirmation message is displayed, false otherwise
     */
    public boolean isConfirmationMessageDisplayedForMaxName() {
        // Placeholder for confirmation message locator (reuse if already defined)
        By confirmationMessage = By.xpath("//div[contains(@class,'confirmation') or contains(text(),'Contact saved')]");
        try {
            return driver.findElement(confirmationMessage).isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }
}

File: src/test/java/pages/ContactCreationPage.ts
// ContactCreationPage.ts
// Page Object for Contact Creation Page (Playwright + TypeScript)
// Follows patterns from ContactCreationPage.java

import { Page, Locator, expect } from '@playwright/test';

export class ContactCreationPage {
  private page: Page;

  // Locators (following descriptive naming and inline pattern from Java PO)
  readonly firstNameInput: Locator;
  readonly lastNameInput: Locator;
  readonly emailAddressInput: Locator;
  readonly emailTypeInput: Locator;
  readonly addEmailButton: Locator;
  readonly categoryDropdown: Locator;
  readonly saveButton: Locator;
  readonly confirmationMessage: Locator;
  // Placeholder for error message locator
  readonly emailFormatErrorMessage: Locator; // TODO: Replace with actual locator

  constructor(page: Page) {
    this.page = page;
    this.firstNameInput = page.locator('[name="first_name"]');
    this.lastNameInput = page.locator('[name="last_name"]');
    this.emailAddressInput = page.locator('input[placeholder="Email address"]');
    this.emailTypeInput = page.locator('input[placeholder="Personal email, Business, Alt..."]');
    this.addEmailButton = page.locator('button.ui.tiny.basic.icon.button i.add.icon');
    this.categoryDropdown = page.locator('div[name="category"]');
    this.saveButton = page.locator('button.ui.linkedin.button i.save.icon');
    this.confirmationMessage = page.locator('div:has-class("confirmation"), div:has-text("Contact saved")');
    this.emailFormatErrorMessage = page.locator('locator("<PLACEHOLDER_email_format_error_message>")'); // TODO: Replace with actual locator
  }

  /**
   * Enter first name
   */
  async enterFirstName(firstName: string): Promise<void> {
    await this.firstNameInput.fill(firstName);
  }

  /**
   * Enter last name
   */
  async enterLastName(lastName: string): Promise<void> {
    await this.lastNameInput.fill(lastName);
  }

  /**
   * Enter email address (can be invalid for negative test)
   */
  async enterEmailAddress(email: string): Promise<void> {
    await this.emailAddressInput.fill(email);
  }

  /**
   * Enter email type (e.g., Personal, Business)
   */
  async enterEmailType(emailType: string): Promise<void> {
    await this.emailTypeInput.fill(emailType);
  }

  /**
   * Click add email button
   */
  async clickAddEmailButton(): Promise<void> {
    await this.addEmailButton.click();
  }

  /**
   * Select category from dropdown
   */
  async selectCategory(category: string): Promise<void> {
    await this.categoryDropdown.click();
    const categoryOption = this.page.locator(`//div[@name='category']//span[text()='${category}']`);
    await categoryOption.click();
  }

  /**
   * Click save button
   */
  async clickSaveButton(): Promise<void> {
    await this.saveButton.click();
  }

  /**
   * Check if confirmation message is displayed (should NOT be for invalid email)
   */
  async isConfirmationMessageDisplayed(): Promise<boolean> {
    return await this.confirmationMessage.isVisible();
  }

  /**
   * Check if email format error message is displayed
   */
  async isEmailFormatErrorMessageDisplayed(): Promise<boolean> {
    // TODO: Replace with actual locator for error message
    return await this.emailFormatErrorMessage.isVisible();
  }

  /**
   * Get email format error message text
   */
  async getEmailFormatErrorMessageText(): Promise<string> {
    // TODO: Replace with actual locator for error message
    return await this.emailFormatErrorMessage.textContent() ?? '';
  }

  /**
   * Complete the negative contact creation flow with invalid email
   */
  async createContactWithInvalidEmail(firstName: string, lastName: string, invalidEmail: string, emailType: string, category: string): Promise<void> {
    await this.enterFirstName(firstName);
    await this.enterLastName(lastName);
    await this.enterEmailAddress(invalidEmail);
    await this.enterEmailType(emailType);
    await this.clickAddEmailButton();
    await this.selectCategory(category);
    await this.clickSaveButton();
  }
}
